contracts/bridge/contracts/common/message/OutboundLane.sol

```diff
@@ -0,0 +1,323 @@
+import "./SourceChain.sol";
+
+// Everything about outgoing messages sending.
+contract OutboundLane is IOutboundLane, MessageVerifier, TargetChain, SourceChain, ReentrancyGuard, AccessControl {
+    event MessageAccepted(uint64 nonce);
+    event MessagesDelivered(uint64 begin, uint64 end, uint256 results);
+    event MessagePruned(uint64 oldest_unpruned_nonce);
+    event MessageFeeIncreased(uint64 nonce, uint256 fee);
+    event RelayerReward(address relayer, uint256 reward);
+
+    bytes32 internal constant OUTBOUND_ROLE = keccak256("OUTBOUND_ROLE");
+    uint64 internal constant MAX_PENDING_MESSAGES = 50;
+    uint64 internal constant MAX_PRUNE_MESSAGES_ATONCE = 10;
+
+    // Outbound lane nonce.
+    struct OutboundLaneNonce {
+        // Nonce of the oldest message that we haven't yet pruned. May point to not-yet-generated
+        // message if all sent messages are already pruned.
+        uint64 oldest_unpruned_nonce;
+        // Nonce of the latest message, received by bridged chain.
+        uint64 latest_received_nonce;
+        // Nonce of the latest message, generated by us.
+        uint64 latest_generated_nonce;
+    }
+
+    /* State */
+
+    OutboundLaneNonce public outboundLaneNonce;
+
+    uint256 public confirmationFee = 0.1 ether; // how to set confirmation_fee
+
+    // nonce => MessageData
+    mapping(uint64 => MessageData) public messages;
+
+    /**
+     * @notice Deploys the OutboundLane contract
+     * @param _lightClientBridge The contract address of on-chain light client
+     * @param _thisChainPosition The thisChainPosition of outbound lane
+     * @param _bridgedChainPosition The bridgedChainPosition of outbound lane
+     * @param _lanePosition The lanePosition of outbound lane
+     * @param _oldest_unpruned_nonce The oldest_unpruned_nonce of outbound lane
+     * @param _latest_received_nonce The latest_received_nonce of outbound lane
+     * @param _latest_generated_nonce The latest_generated_nonce of outbound lane
+     */
+    constructor(
+        address _lightClientBridge,
+        uint32 _thisChainPosition,
+        uint32 _bridgedChainPosition,
+        uint32 _lanePosition,
+        uint64 _oldest_unpruned_nonce,
+        uint64 _latest_received_nonce,
+        uint64 _latest_generated_nonce
+    ) public MessageVerifier(_lightClientBridge, _thisChainPosition, _bridgedChainPosition, _lanePosition) {
+        outboundLaneNonce = OutboundLaneNonce(_oldest_unpruned_nonce, _latest_received_nonce, _latest_generated_nonce);
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    // Send message over lane.
+    function send_message(address targetContract, bytes calldata encoded) external payable override nonReentrant returns (uint64) {
+        require(hasRole(OUTBOUND_ROLE, msg.sender), "Lane: NotAuthorized");
+        require(outboundLaneNonce.latest_generated_nonce - outboundLaneNonce.latest_received_nonce <= MAX_PENDING_MESSAGES, "Lane: TooManyPendingMessages");
```

Comment:
`MAX_PENDING_MESSAGES` might not be approximate here, since it it only storing message data, and we can limit buffer using `MAX_UNCONFIRMED_MESSAGES` when delivering messages on target chain(relayers will help handle this.).

Beside, the side effect is it will limit the usage from client, and cause bad UX.